File: 第二阶段/2.7-functools与itertools/FunctoolsItertools.cs
// C# LINQ + Func 示例
// 对应文章：2.7 functools与itertools
//
// C# 老兵的 Python 修炼手册：
// Python 的 functools + itertools 组合拳，在 C# 中主要靠 LINQ + Func 闭包来接招。
// 有些功能 C# 更强（如真正的泛型、类型安全），有些 Python 更爽（如排列组合内置支持）。

using System;
using System.Collections.Generic;
using System.Linq;
using System.Runtime.CompilerServices;

public class FunctoolsItertoolsDemo
{
    // ================================================================
    //                       functools 对应部分
    // ================================================================

    // 1. Aggregate（等价于 Python 的 reduce）
    // Python: reduce(lambda acc, x: acc + x, numbers)
    // C#:     numbers.Aggregate((acc, x) => acc + x)
    static void ReduceExample()
    {
        var numbers = new[] { 1, 2, 3, 4, 5 };
        int total = numbers.Aggregate((acc, x) => acc + x);
        int productResult = numbers.Aggregate(1, (acc, x) => acc * x);
        Console.WriteLine($"  求和: {total}");      // 15
        Console.WriteLine($"  求积: {productResult}"); // 120
    }

    // 2. Memoize（等价于 Python 的 lru_cache）
    // Python: @lru_cache(maxsize=128)
    // C#:     手动维护 Dictionary（或 .NET 8+ 的 [Memoize] 特性）
    private static Dictionary<int, long> _fibCache = new();
    static long Fibonacci(int n)
    {
        if (n < 2) return n;
        if (_fibCache.TryGetValue(n, out var cached))
            return cached;
        long result = Fibonacci(n - 1) + Fibonacci(n - 2);
        _fibCache[n] = result;
        return result;
    }

    // 3. Func 偏函数（等价于 Python 的 partial）
    // Python: square = partial(power, exponent=2)
    // C#:     var square = Power(2);  —— 用闭包实现
    static Func<int, int> Power(int exponent)
    {
        return baseValue => (int)Math.Pow(baseValue, exponent);
    }

    // 4. Concat（等价于 Python 的 chain）
    // Python: chain(list1, list2, list3)
    // C#:     list1.Cast<object>().Concat(list2.Cast<object>())...
    //         （因为 C# 是强类型的，不同类型合并需要统一类型）
    static void ChainExample()
    {
        var list1 = new[] { 1, 2, 3 };
        var list2 = new[] { "a", "b", "c" };
        var list3 = new[] { 100, 200 };
        // C# 中类型需一致，这里用 object 演示
        var combined = list1.Cast<object>()
            .Concat(list2.Cast<object>())
            .Concat(list3.Cast<object>())
            .ToList();
        Console.WriteLine($"  chain: [{string.Join(", ", combined)}]");
    }

    // 5. Take（等价于 Python 的 islice）
    // Python: islice(count(1), 5)
    // C#:     Enumerable.Range(1, 5)
    static void IsliceExample()
    {
        var firstFive = Enumerable.Range(1, 5).ToList();
        Console.WriteLine($"  前5个自然数: [{string.Join(", ", firstFive)}]");
    }

    // 6. Where + 索引（等价于 Python 的 compress）
    // Python: compress(data, mask)
    // C#:     data.Where((_, i) => mask[i])
    static void CompressExample()
    {
        var data = new[] { "A", "B", "C", "D", "E" };
        var mask = new[] { true, false, true, false, true };
        var filtered = data.Where((_, i) => mask[i]).ToList();
        Console.WriteLine($"  compress: [{string.Join(", ", filtered)}]");
    }

    // 7. GroupBy（等价于 Python 的 groupby）
    // Python: groupby(data, key=itemgetter(0)) —— 要求预排序！
    // C#:     data.GroupBy(x => x.Item1) —— 不要求预排序，更方便
    static void GroupByExample()
    {
        var data = new (string, int)[]
        {
            ("A", 1), ("A", 2), ("B", 3), ("B", 4), ("A", 5)
        };
        // C# GroupBy 不要求预排序，这是比 Python 更强的地方
        foreach (var group in data.GroupBy(x => x.Item1))
        {
            Console.WriteLine($"  组 {group.Key}: [{string.Join(", ", group)}]");
        }
    }

    // 8. Cartesian Product（等价于 Python 的 product）
    // Python: product(suits, ranks)
    // C#:     suits.SelectMany(s => ranks, (s, r) => (s, r))
    static void ProductExample()
    {
        var suits = new[] { "♠", "♥" };
        var ranks = new[] { "A", "K", "Q" };
        var cards = suits.SelectMany(s => ranks, (s, r) => $"({s},{r})").ToList();
        Console.WriteLine($"  扑克牌: [{string.Join(", ", cards)}]");
    }

    // 9. Select + 索引（等价于 Python 的 starmap）
    // Python: starmap(mul, pairs)
    // C#:     pairs.Select(p => p.Item1 * p.Item2)
    static void StarmapExample()
    {
        var pairs = new[] { (2, 3), (4, 5), (6, 7) };
        var results = pairs.Select(p => p.Item1 * p.Item2).ToList();
        Console.WriteLine($"  starmap mul: [{string.Join(", ", results)}]");
    }

    // 10. TakeWhile / SkipWhile（等价于 Python 的 takewhile / dropwhile）
    // Python: takewhile(lambda x: x < 5, numbers)
    // C#:     numbers.TakeWhile(n => n < 5)
    // Python: dropwhile(lambda x: x < 5, numbers)
    // C#:     numbers.SkipWhile(n => n < 5)
    static void TakeWhileDropWhileExample()
    {
        var numbers = new[] { 1, 3, 5, 2, 4, 6, 8 };
        var taken = numbers.TakeWhile(n => n < 5).ToList();
        var dropped = numbers.SkipWhile(n => n < 5).ToList();
        Console.WriteLine($"  原始数据: [{string.Join(", ", numbers)}]");
        Console.WriteLine($"  TakeWhile(x < 5): [{string.Join(", ", taken)}]");
        Console.WriteLine($"  SkipWhile(x < 5): [{string.Join(", ", dropped)}]");
        Console.WriteLine($"  C# 的 TakeWhile/SkipWhile 和 Python 几乎一模一样！");
    }

    // 11. Zip（等价于 Python 的 zip_longest 的简化版）
    // Python: zip_longest(names, scores, fillvalue="(缺考)")
    // C#:     names.Zip(scores, (n, s) => (n, s)) —— 默认以短的为准
    //         要实现 fillvalue 效果需要手动处理
    static void ZipLongestExample()
    {
        var names = new[] { "Alice", "Bob", "Charlie", "Diana" };
        var scores = new[] { 95, 87, 91 };
        // C# 的 Zip 默认以短的为准
        var normalZip = names.Zip(scores, (n, s) => $"({n},{s})").ToList();
        Console.WriteLine($"  普通 Zip: [{string.Join(", ", normalZip)}]");
        // 模拟 zip_longest（C# 没有内置的 zip_longest）
        var longZip = names.Select((n, i) =>
            i < scores.Length ? $"({n},{scores[i]})" : $"({n},(缺考))").ToList();
        Console.WriteLine($"  模拟 zip_longest: [{string.Join(", ", longZip)}]");
        Console.WriteLine($"  C# 没有内置的 zip_longest，需要手动处理");
    }

    // 12. 排列组合（等价于 Python 的 permutations / combinations）
    // Python: list(permutations(items, 2)) —— 内置！超方便！
    // C#:     需要自己写递归实现，或用 MoreLINQ 等第三方库
    // 这是 Python itertools 最让 C# 羡慕的功能之一！
    static List<List<T>> Permutations<T>(List<T> items, int k)
    {
        var result = new List<List<T>>();
        PermuteHelper(items, k, new HashSet<int>(), new List<T>(), result);
        return result;
    }

    static void PermuteHelper<T>(List<T> items, int k,
        HashSet<int> used, List<T> current, List<List<T>> result)
    {
        if (current.Count == k)
        {
            result.Add(new List<T>(current));
            return;
        }
        for (int i = 0; i < items.Count; i++)
        {
            if (used.Contains(i)) continue;
            used.Add(i);
            current.Add(items[i]);
            PermuteHelper(items, k, used, current, result);
            current.RemoveAt(current.Count - 1);
            used.Remove(i);
        }
    }

    static List<List<T>> Combinations<T>(List<T> items, int k)
    {
        var result = new List<List<T>>();
        CombineHelper(items, k, 0, new List<T>(), result);
        return result;
    }

    static void CombineHelper<T>(List<T> items, int k, int start,
        List<T> current, List<List<T>> result)
    {
        if (current.Count == k)
        {
            result.Add(new List<T>(current));
            return;
        }
        for (int i = start; i < items.Count; i++)
        {
            current.Add(items[i]);
            CombineHelper(items, k, i + 1, current, result);
            current.RemoveAt(current.Count - 1);
        }
    }

    static void PermutationCombinationExample()
    {
        var items = new List<string> { "A", "B", "C" };

        var perms = Permutations(items, 2);
        var permStrings = perms.Select(p => $"({string.Join(",", p)})").ToList();
        Console.WriteLine($"  permutations(ABC, 2): [{string.Join(", ", permStrings)}]");
        Console.WriteLine($"  共 {perms.Count} 种排列（顺序不同算不同）");

        var combos = Combinations(items, 2);
        var comboStrings = combos.Select(c => $"({string.Join(",", c)})").ToList();
        Console.WriteLine($"  combinations(ABC, 2): [{string.Join(", ", comboStrings)}]");
        Console.WriteLine($"  共 {combos.Count} 种组合（顺序相同算同一种）");

        var fullPerms = Permutations(items, 3);
        Console.WriteLine($"  全排列: {fullPerms.Count} = 3! 种");
        Console.WriteLine($"  C# 没有内置排列组合，需要自己写递归或用 MoreLINQ 等库");
    }

    static void Main()
    {
        Console.WriteLine("===== functools & itertools 对比 =====\n");

        Console.WriteLine("--- 1. reduce vs Aggregate ---");
        ReduceExample();

        Console.WriteLine("\n--- 2. lru_cache vs 手动缓存 ---");
        Console.WriteLine($"  fib(30) = {Fibonacci(30)}");

        Console.WriteLine("\n--- 3. partial vs Func 闭包 ---");
        var square = Power(2);
        var cube = Power(3);
        Console.WriteLine($"  平方: {square(5)}");  // 25
        Console.WriteLine($"  立方: {cube(5)}");    // 125

        Console.WriteLine("\n--- 4. chain vs Concat ---");
        ChainExample();

        Console.WriteLine("\n--- 5. islice vs Take ---");
        IsliceExample();

        Console.WriteLine("\n--- 6. compress vs Where ---");
        CompressExample();

        Console.WriteLine("\n--- 7. groupby vs GroupBy ---");
        GroupByExample();

        Console.WriteLine("\n--- 8. product vs SelectMany ---");
        ProductExample();

        Console.WriteLine("\n--- 9. starmap vs Select ---");
        StarmapExample();

        Console.WriteLine("\n--- 10. takewhile/dropwhile vs TakeWhile/SkipWhile ---");
        TakeWhileDropWhileExample();

        Console.WriteLine("\n--- 11. zip_longest vs Zip ---");
        ZipLongestExample();

        Console.WriteLine("\n--- 12. permutations/combinations vs 手写递归 ---");
        PermutationCombinationExample();

        Console.WriteLine("\n===== 总结: LINQ 已经很强了，但 Python 的排列组合真的很香！=====");
    }
}

File: 第二阶段/2.7-functools与itertools/functools_itertools.py
# Python functools 与 itertools 示例
# 对应文章：2.7 functools与itertools
#
# C# 老兵的 Python 修炼手册：
# functools 是 Python 的"函数工具箱"，相当于 C# 的扩展方法 + 特性；
# itertools 是 Python 的"迭代器军火库"，相当于 C# 的 LINQ + 一些黑科技。

from functools import (
    reduce, lru_cache, partial, singledispatch, cache
)
from itertools import (
    count, cycle, repeat, chain, islice,
    compress, filterfalse, starmap, groupby, product,
    takewhile, dropwhile, zip_longest,
    permutations, combinations
)
from operator import mul, itemgetter

# ================================================================
#                       functools 示例
# ================================================================

print("=" * 60)
print("               functools 模块 —— 函数的瑞士军刀")
print("=" * 60)

# 1. reduce —— 累积计算（C# 中用 LINQ 的 Aggregate）
# C# 对应: numbers.Aggregate((acc, x) => acc + x)
numbers = [1, 2, 3, 4, 5]
total = reduce(lambda acc, x: acc + x, numbers)
product_result = reduce(mul, numbers)
print(f"\n--- 1. reduce（累积计算）---")
print(f"  求和: {total}")        # 15
print(f"  求积: {product_result}")  # 120

# 2. lru_cache —— LRU 缓存（C# 需手动用 Dictionary 实现）
# C# 对应: 手动维护 Dictionary<int, long> 做缓存，或用 [Memoize] 特性（.NET 8+）
@lru_cache(maxsize=128)
def fibonacci(n):
    """经典的斐波那契 —— 有缓存后从 O(2^n) 变成 O(n)"""
    if n < 2:
        return n
    return fibonacci(n - 1) + fibonacci(n - 2)

print(f"\n--- 2. lru_cache（LRU 缓存）---")
print(f"  fib(30) = {fibonacci(30)}")
print(f"  缓存命中: {fibonacci.cache_info()}")

# 3. cache —— Python 3.9+ 的无限缓存（比 lru_cache 更简洁）
# C# 对应: 没有内置的无限缓存，需要自己封装
#           [Memoize] 特性（.NET 8+）是类似的思路，但没有大小限制管理
# cache = lru_cache(maxsize=None)  # 本质上就是这样！

@cache
def factorial(n):
    """用 @cache 装饰的阶乘 —— 比 @lru_cache 写起来更爽"""
    if n <= 1:
        return 1
    return n * factorial(n - 1)

print(f"\n--- 3. cache（无限缓存，Python 3.9+）---")
print(f"  10! = {factorial(10)}")
print(f"  20! = {factorial(20)}")
print(f"  特点: 不需要指定 maxsize，永远不会淘汰缓存")
print(f"  对比 @lru_cache: cache = lru_cache(maxsize=None)，但写起来更短更爽")
print(f"  C# 对比: .NET 8 的 [Memoize] 特性有点类似，但目前还不支持异步")

# 4. partial —— 偏函数（固定部分参数）
# C# 对应: Func<T> 闭包，或者带默认参数的方法
def power(base, exponent):
    return base ** exponent

square = partial(power, exponent=2)
cube = partial(power, exponent=3)
print(f"\n--- 4. partial（偏函数）---")
print(f"  平方: {square(5)}")   # 25
print(f"  立方: {cube(5)}")     # 125

# 5. singledispatch —— 单分派泛型函数（C# 有真正的泛型！）
# C# 对应: 方法重载 / 模式匹配（switch + when）
@singledispatch
def to_string(value):
    return str(value)

@to_string.register(int)
def _(value):
    return f"整数: {value}"

@to_string.register(list)
def _(value):
    return f"列表(长度{len(value)}): {value}"

@to_string.register(dict)
def _(value):
    return f"字典({len(value)}个键): {list(value.keys())}"

print(f"\n--- 5. singledispatch（单分派泛型）---")
print(f"  {to_string(42)}")           # 整数: 42
print(f"  {to_string([1, 2, 3])}")   # 列表(长度3): [1, 2, 3]
print(f"  {to_string({'a': 1})}")    # 字典(1个键): ['a']
print(f"  C# 对比: 直接用方法重载就行，C# 的类型系统更强")

# ================================================================
#                       itertools 示例
# ================================================================

print("\n" + "=" * 60)
print("               itertools 模块 —— 迭代器的军火库")
print("=" * 60)

# 6. chain —— 串联多个可迭代对象（C# 用 Concat）
# C# 对应: list1.Concat(list2).Concat(list3)
list1 = [1, 2, 3]
list2 = ["a", "b", "c"]
list3 = [100, 200]
combined = list(chain(list1, list2, list3))
print(f"\n--- 6. chain（串联）---")
print(f"  结果: {combined}")

# 7. islice —— 切片迭代器（C# 用 Take/Skip）
# C# 对应: Enumerable.Range(1, 5)
first_five = list(islice(count(1), 5))
print(f"\n--- 7. islice（切片）---")
print(f"  前5个自然数: {first_five}")

# 8. compress —— 按掩码过滤（C# 用 Where + 索引）
# C# 对应: data.Where((item, index) => mask[index])
data = ["A", "B", "C", "D", "E"]
mask = [1, 0, 1, 0, 1]
filtered = list(compress(data, mask))
print(f"\n--- 8. compress（掩码过滤）---")
print(f"  结果: {filtered}")  # ['A', 'C', 'E']

# 9. groupby —— 分组（C# 有 GroupBy，但需要预排序！）
# C# 对应: data.GroupBy(x => x.Item1)，但 C# 的 GroupBy 不要求预排序
data = [("A", 1), ("A", 2), ("B", 3), ("B", 4), ("A", 5)]
data.sort(key=itemgetter(0))
print(f"\n--- 9. groupby（分组）---")
for key, group in groupby(data, key=itemgetter(0)):
    print(f"  组 {key}: {list(group)}")
print(f"  注意: Python 的 groupby 要求数据先排序！C# 的 GroupBy 不需要")

# 10. product —— 笛卡尔积（C# 用 SelectMany）
# C# 对应: suits.SelectMany(s => ranks, (s, r) => (s, r))
suits = ["S", "H"]
ranks = ["A", "K", "Q"]
cards = [(s, r) for s, r in product(suits, ranks)]
print(f"\n--- 10. product（笛卡尔积）---")
print(f"  扑克牌: {cards}")

# 11. starmap —— 星号映射（C# 用 Select）
# C# 对应: pairs.Select(p => p.Item1 * p.Item2)
pairs = [(2, 3), (4, 5), (6, 7)]
results = list(starmap(mul, pairs))
print(f"\n--- 11. starmap（星号映射）---")
print(f"  结果: {results}")  # [6, 20, 42]

# 12. takewhile / dropwhile —— 条件截断（C# 用 TakeWhile / SkipWhile）
# C# 对应: numbers.TakeWhile(n => n < 4)
#           numbers.SkipWhile(n => n % 2 == 0)
# 这两个是 Python 迭代器的"杀手级"功能，用起来特别优雅！
numbers_tw = [1, 3, 5, 2, 4, 6, 8]
taken = list(takewhile(lambda x: x < 5, numbers_tw))
dropped = list(dropwhile(lambda x: x < 5, numbers_tw))
print(f"\n--- 12. takewhile / dropwhile（条件截断）---")
print(f"  原始数据: {numbers_tw}")
print(f"  takewhile(x < 5): {taken}")   # [1, 3] —— 遇到 5 就停
print(f"  dropwhile(x < 5): {dropped}") # [5, 2, 4, 6, 8] —— 跳过小于5的，之后全保留
print(f"  C# 对比: TakeWhile / SkipWhile，几乎一模一样")

# 13. zip_longest —— 不等长拉链（C# 用 Zip + 第三个参数）
# C# 对应: list1.Zip(list2, (a, b) => (a, b)) 但 C# 的 Zip 默认以短的为准
#           要实现 zip_longest 效果需要用更多代码
names = ["Alice", "Bob", "Charlie", "Diana"]
scores = [95, 87, 91]
# 普通 zip 会丢掉 Diana
normal_zip = list(zip(names, scores))
# zip_longest 用 None 填充
full_zip = list(zip_longest(names, scores, fillvalue="(缺考)"))
print(f"\n--- 13. zip_longest（不等长拉链）---")
print(f"  普通 zip:  {normal_zip}")
print(f"  zip_longest: {full_zip}")
print(f"  C# 对比: C# 的 Zip 默认以短的为准，要填充需要额外处理")

# 14. permutations / combinations —— 排列组合（C# 需要手写或用第三方库）
# C# 没有内置的排列组合函数，通常需要自己写递归或用 NuGet 包
# 这是 Python 的 itertools 最让人羡慕的功能之一！
print(f"\n--- 14. permutations / combinations（排列组合）---")
items = ["A", "B", "C"]

# 排列：考虑顺序，从3个中取2个
perms = list(permutations(items, 2))
print(f"  permutations(ABC, 2) = {perms}")
print(f"  共 {len(perms)} 种排列（顺序不同算不同）")

# 组合：不考虑顺序，从3个中取2个
combos = list(combinations(items, 2))
print(f"  combinations(ABC, 2) = {combos}")
print(f"  共 {len(combos)} 种组合（顺序相同算同一种）")

# 全排列
full_perms = list(permutations(items))
print(f"  全排列: {full_perms}")
print(f"  共 {len(full_perms)} = 3! 种")
print(f"  C# 对比: C# 没有内置排列组合，需要自己写递归或用 MoreLINQ 等库")

print("\n" + "=" * 60)
print("  总结: functools 是函数的增强工具箱，itertools 是迭代器的军火库")
print("  C# 开发者的感受: LINQ 已经很强了，但 itertools 的排列组合真的很香！")
print("=" * 60)
